File: python/bot/utils/automod_thresholds.py
"""Automod threshold validation and display helpers."""

from __future__ import annotations

import re
from typing import Any

from bot.constants import AUTOMOD_THRESHOLD_DEFAULTS, CAPS_MIN_LETTERS
from bot.database.repositories.guild_settings import GuildSettings
from bot.utils.time import format_duration, parse_duration

_THRESHOLD_RANGES: dict[str, tuple[int, int]] = {
    "caps_threshold": (50, 100),
    "spam_threshold": (3, 20),
    "spam_interval_ms": (1000, 60_000),
    "mention_threshold": (2, 50),
}


def _parse_int(raw: str) -> int | None:
    trimmed = raw.strip()
    if re.fullmatch(r"\d+", trimmed):
        return int(trimmed)
    return None


def validate_caps_threshold_input(raw: str) -> dict[str, Any]:
    value = _parse_int(raw)
    low, high = _THRESHOLD_RANGES["caps_threshold"]
    if value is None:
        return {"ok": False, "error": "Caps threshold must be a whole number percentage."}
    if value < low or value > high:
        return {"ok": False, "error": f"Caps threshold must be between {low} and {high}."}
    return {"ok": True, "value": value}


def validate_spam_count_input(raw: str) -> dict[str, Any]:
    value = _parse_int(raw)
    low, high = _THRESHOLD_RANGES["spam_threshold"]
    if value is None:
        return {"ok": False, "error": "Spam count must be a whole number."}
    if value < low or value > high:
        return {"ok": False, "error": f"Spam count must be between {low} and {high} messages."}
    return {"ok": True, "value": value}


def validate_spam_window_input(raw: str) -> dict[str, Any]:
    trimmed = raw.strip()
    if not trimmed:
        return {"ok": False, "error": "Provide a duration such as `5s` or `1m`."}
    ms_value = _parse_int(trimmed) or parse_duration(trimmed)
    if ms_value is None:
        return {"ok": False, "error": "Spam window must be a duration such as `5s` or `1m`."}
    low, high = _THRESHOLD_RANGES["spam_interval_ms"]
    if ms_value < low or ms_value > high:
        return {"ok": False, "error": "Spam window must be between 1 second and 60 seconds."}
    return {"ok": True, "value": int(ms_value)}


def validate_mention_threshold_input(raw: str) -> dict[str, Any]:
    value = _parse_int(raw)
    low, high = _THRESHOLD_RANGES["mention_threshold"]
    if value is None:
        return {"ok": False, "error": "Mention threshold must be a whole number."}
    if value < low or value > high:
        return {"ok": False, "error": f"Mention threshold must be between {low} and {high}."}
    return {"ok": True, "value": value}


def format_spam_window(ms_value: int) -> str:
    return format_duration(ms_value)


def format_threshold_show(settings: GuildSettings, *, module_disabled: bool = False) -> str:
    inactive = " (saved, inactive — Automod module disabled)" if module_disabled else ""
    lines = [
        (
            f"**Caps:** {settings.caps_threshold}% at {CAPS_MIN_LETTERS}+ letters "
            f"— anti-caps {'enabled' if settings.anti_caps else 'disabled'}{inactive}"
        ),
        (
            f"**Spam:** {settings.spam_threshold} messages within "
            f"{format_spam_window(settings.spam_interval_ms)} "
            f"— anti-spam {'enabled' if settings.anti_spam else 'disabled'}{inactive}"
        ),
        (
            f"**Mentions:** {settings.mention_threshold} user/role mentions per message "
            f"(@everyone/@here always flagged) "
            f"— anti-mention {'enabled' if settings.anti_mention else 'disabled'}{inactive}"
        ),
        "",
        "_Minimum caps length is fixed at 8 letters (not configurable)._",
        "_Use `?automod threshold reset <caps|spam|mentions|all>` to restore defaults._",
    ]
    return "\n".join(lines)


def get_threshold_reset_updates(target: str) -> dict[str, int] | None:
    normalized = target.lower()
    if normalized == "caps":
        return {"caps_threshold": AUTOMOD_THRESHOLD_DEFAULTS["caps_threshold"]}
    if normalized == "spam":
        return {
            "spam_threshold": AUTOMOD_THRESHOLD_DEFAULTS["spam_threshold"],
            "spam_interval_ms": AUTOMOD_THRESHOLD_DEFAULTS["spam_interval_ms"],
        }
    if normalized == "mentions":
        return {"mention_threshold": AUTOMOD_THRESHOLD_DEFAULTS["mention_threshold"]}
    if normalized == "all":
        return dict(AUTOMOD_THRESHOLD_DEFAULTS)
    return None

File: node-bun/src/utils/automodThresholds.js
import { formatDuration } from './time.js';

export const CAPS_MIN_LETTERS = 8;

export const AUTOMOD_THRESHOLD_DEFAULTS = {
  caps_threshold: 70,
  spam_threshold: 5,
  spam_interval: 5000,
  mention_threshold: 5,
};

export const THRESHOLD_KEYS = Object.keys(AUTOMOD_THRESHOLD_DEFAULTS);

const INTEGER_KEYS = new Set(['caps_threshold', 'spam_threshold', 'mention_threshold']);

export const THRESHOLD_RANGES = {
  caps_threshold: { min: 50, max: 100 },
  spam_threshold: { min: 3, max: 20 },
  spam_interval: { min: 1000, max: 60000 },
  mention_threshold: { min: 2, max: 50 },
};

function parseInteger(raw) {
  if (typeof raw === 'number' && Number.isInteger(raw)) return raw;
  if (typeof raw === 'string' && /^\d+$/.test(raw.trim())) return parseInt(raw.trim(), 10);
  return null;
}

function parseMs(raw) {
  if (typeof raw === 'number' && Number.isFinite(raw)) return Math.round(raw);
  if (typeof raw === 'string' && /^\d+$/.test(raw.trim())) return parseInt(raw.trim(), 10);
  return null;
}

export function coerceThresholdValue(key, raw, defaultValue = AUTOMOD_THRESHOLD_DEFAULTS[key]) {
  const range = THRESHOLD_RANGES[key];
  if (!range) return { value: defaultValue, replaced: false };

  if (raw === undefined || raw === null) {
    return { value: defaultValue, replaced: false };
  }

  let parsed;
  if (key === 'spam_interval') {
    parsed = parseMs(raw);
  } else {
    parsed = parseInteger(raw);
    if (parsed !== null && !INTEGER_KEYS.has(key)) {
      parsed = null;
    }
  }

  if (parsed === null || !Number.isFinite(parsed)) {
    return { value: defaultValue, replaced: true };
  }

  if (INTEGER_KEYS.has(key) && !Number.isInteger(parsed)) {
    return { value: defaultValue, replaced: true };
  }

  if (parsed < range.min || parsed > range.max) {
    return { value: defaultValue, replaced: true };
  }

  return { value: parsed, replaced: false };
}

export function normalizeAutomodThresholds(settings, { log } = {}) {
  const normalized = { ...settings };
  for (const key of THRESHOLD_KEYS) {
    const { value, replaced } = coerceThresholdValue(key, settings[key], AUTOMOD_THRESHOLD_DEFAULTS[key]);
    if (replaced && settings[key] !== undefined && settings[key] !== value && log) {
      log(`[automod] Invalid ${key}=${JSON.stringify(settings[key])}; using ${value}`);
    }
    normalized[key] = value;
  }
  return normalized;
}

export function resolveAutomodThresholds(settings) {
  return normalizeAutomodThresholds(settings);
}

export function validateCapsThresholdInput(input) {
  const trimmed = input?.trim();
  if (!trimmed || !/^\d+$/.test(trimmed)) {
    return { ok: false, error: 'Caps threshold must be a whole number percentage.' };
  }
  const value = parseInt(trimmed, 10);
  const range = THRESHOLD_RANGES.caps_threshold;
  if (value < range.min || value > range.max) {
    return { ok: false, error: `Caps threshold must be between ${range.min} and ${range.max}.` };
  }
  return { ok: true, value };
}

export function validateSpamCountInput(input) {
  const trimmed = input?.trim();
  if (!trimmed || !/^\d+$/.test(trimmed)) {
    return { ok: false, error: 'Spam count must be a whole number.' };
  }
  const value = parseInt(trimmed, 10);
  const range = THRESHOLD_RANGES.spam_threshold;
  if (value < range.min || value > range.max) {
    return { ok: false, error: `Spam count must be between ${range.min} and ${range.max} messages.` };
  }
  return { ok: true, value };
}

export function validateSpamWindowInput(input, parseDuration) {
  const trimmed = input?.trim();
  if (!trimmed) {
    return { ok: false, error: 'Provide a duration such as `5s` or `1m`.' };
  }

  let msValue;
  if (/^\d+$/.test(trimmed)) {
    msValue = parseInt(trimmed, 10);
  } else {
    msValue = parseDuration(trimmed);
  }

  if (!msValue || !Number.isFinite(msValue)) {
    return { ok: false, error: 'Spam window must be a duration such as `5s` or `1m`.' };
  }

  const range = THRESHOLD_RANGES.spam_interval;
  if (msValue < range.min || msValue > range.max) {
    return { ok: false, error: `Spam window must be between 1 second and 60 seconds.` };
  }

  return { ok: true, value: Math.round(msValue) };
}

export function validateMentionThresholdInput(input) {
  const trimmed = input?.trim();
  if (!trimmed || !/^\d+$/.test(trimmed)) {
    return { ok: false, error: 'Mention threshold must be a whole number.' };
  }
  const value = parseInt(trimmed, 10);
  const range = THRESHOLD_RANGES.mention_threshold;
  if (value < range.min || value > range.max) {
    return { ok: false, error: `Mention threshold must be between ${range.min} and ${range.max}.` };
  }
  return { ok: true, value };
}

export function capsPercentage(content) {
  const letters = content.replace(/[^a-zA-Z]/g, '');
  if (letters.length < CAPS_MIN_LETTERS) return 0;
  const caps = letters.replace(/[^A-Z]/g, '').length;
  return (caps / letters.length) * 100;
}

export function countMentionTargets(message) {
  return message.mentions.users.size + message.mentions.roles.size;
}

export function isMassMention(message, mentionThreshold) {
  if (message.mentions.everyone) return true;
  return countMentionTargets(message) >= mentionThreshold;
}

export function formatSpamWindow(msValue) {
  return formatDuration(msValue);
}

export function formatThresholdShow(settings, { moduleDisabled = false } = {}) {
  const thresholds = resolveAutomodThresholds(settings);
  const inactive = moduleDisabled ? ' (saved, inactive — Automod module disabled)' : '';

  const lines = [
    `**Caps:** ${thresholds.caps_threshold}% at ${CAPS_MIN_LETTERS}+ letters — anti-caps ${settings.anti_caps ? 'enabled' : 'disabled'}${inactive}`,
    `**Spam:** ${thresholds.spam_threshold} messages within ${formatSpamWindow(thresholds.spam_interval)} — anti-spam ${settings.anti_spam ? 'enabled' : 'disabled'}${inactive}`,
    `**Mentions:** ${thresholds.mention_threshold} user/role mentions per message (@everyone/@here always flagged) — anti-mention ${settings.anti_mention ? 'enabled' : 'disabled'}${inactive}`,
    '',
    '_Minimum caps length is fixed at 8 letters (not configurable)._',
    '_Use `?automod threshold reset <caps|spam|mentions|all>` to restore defaults._',
  ];

  return lines.join('\n');
}

export function getThresholdResetUpdates(target) {
  switch (target) {
    case 'caps':
      return { caps_threshold: AUTOMOD_THRESHOLD_DEFAULTS.caps_threshold };
    case 'spam':
      return {
        spam_threshold: AUTOMOD_THRESHOLD_DEFAULTS.spam_threshold,
        spam_interval: AUTOMOD_THRESHOLD_DEFAULTS.spam_interval,
      };
    case 'mentions':
      return { mention_threshold: AUTOMOD_THRESHOLD_DEFAULTS.mention_threshold };
    case 'all':
      return { ...AUTOMOD_THRESHOLD_DEFAULTS };
    default:
      return null;
  }
}
